localstack/services/s3/website_hosting.py

```diff
@@ -0,0 +1,376 @@
+    """
+    parsed_url = urlparse(request.url)
+    redirect_to = request.url
+    redirect = routing_rule["Redirect"]
+    if host_name := redirect.get("HostName"):
+        redirect_to = redirect_to.replace(parsed_url.netloc, host_name)
+    if protocol := redirect.get("Protocol"):
+        redirect_to = redirect_to.replace(parsed_url.scheme, protocol)
+    if redirect_to_key := redirect.get("ReplaceKeyWith"):
+        redirect_to = redirect_to.replace(parsed_url.path, f"/{redirect_to_key}")
+    elif new_key_prefix := redirect.get("ReplaceKeyPrefixWith"):
+        # TODO test if KeyPrefixEquals is not supplied???
+        matched_prefix = routing_rule["Condition"].get("KeyPrefixEquals")
+        redirect_to = redirect_to.replace(matched_prefix, new_key_prefix)
+
+    return Response(
+        "", headers={"Location": redirect_to}, status=redirect.get("HttpRedirectCode", 301)
+    )
+
+
+def get_error_document(
+    website_config: WebsiteConfiguration, bucket: FakeBucket, missing_key: ObjectKey
+) -> Response:
+    """
+    Either tries to get the
+    https://docs.aws.amazon.com/AmazonS3/latest/userguide/CustomErrorDocSupport.html
+    :param website_config: the bucket WebsiteConfiguration
+    :param bucket: the bucket object from moto
+    :param missing_key: the missing key not found in the bucket
+    :return:
+    """
+    headers = {}
+    if error_document := website_config.get("ErrorDocument"):
+        # if an error document is configured, try to fetch the key
+        error_key = error_document["Key"]
+        key = get_key_from_moto_bucket(bucket, error_key)
+        if key:
+            # if the key is found, return the key, or if that key has a redirect, return a redirect
+            error_body = key.value
+            if key.website_redirect_location:
+                headers["Location"] = key.website_redirect_location
+                return Response("", status=301, headers=headers)
+
+            headers = get_response_headers_from_key(key)
+            return Response(error_body, status=404, headers=headers)
+        else:
+            ex = NoSuchKeyFromErrorDocument("The specified key does not exist.")
+            ex.Key = missing_key
+            ex.ErrorDocumentKey = error_key
+            raise ex
+
+    else:
+        ex = NoSuchKey("The specified key does not exist.")
+        ex.Key = missing_key
+        raise ex
+
+
+def check_if_headers(headers: Headers, key: FakeKey) -> bool:
+    # TODO: add other conditions here If-Modified-Since, etc etc
+    if "if-none-match" in headers and key.etag and key.etag in headers["if-none-match"]:
+        return True
+
+
+def register_website_hosting_routes(router: Router[Handler]):
+    """
+    Registers the S3 website hosting handlers into the given router.
+
+    :param router: the router to add the handlers into.
+    """
+    router.add(
+        path="/",
+        host=STATIC_WEBSITE_HOST_REGEX,
+        endpoint=website_handler,
+        methods=["GET"],
```

Comment:
have you checked if you make other HTTP (POST) requests to websites? does that work? what does aws return otherwise, 404?

Reply:
I'll check 👍

Reply:
So it returns 405, with some explanation. I've created a template for 405 errors, snapshotted it and tested it. Thanks for the pointer! 👍